src/i_midimusic.c

```diff
@@ -1383,287 +1148,261 @@ static boolean IsRPGLoop(void)
     return (num_rpg_events == 1 && num_emidi_events == 0);
 }
 
-// Fills the output buffer with events from the current song and then streams it
-// out. Call this function from the MIDI thread only, with exclusive access to
-// shared resources.
+// Get the next event from the MIDI file, process it or return if the delta
+// time is > 0. Call this function from the MIDI thread only, with exclusive
+// access to shared resources.
 
-static void FillBuffer(void)
+static midi_state_t NextEvent(midi_event_t **event, midi_track_t **track,
+                              int64_t *wait_time)
 {
-    unsigned int i;
-    int num_events;
+    midi_event_t *local_event = NULL;
+    midi_track_t *local_track = NULL;
+    unsigned int min_time = UINT_MAX;
+    unsigned int delta_time;
 
-    for (num_events = 0; num_events < STREAM_MAX_EVENTS;)
+    // Find next event across all tracks.
+    for (int i = 0; i < song.num_tracks; ++i)
     {
-        midi_event_t *event = NULL;
-        win_midi_track_t *track = NULL;
-        unsigned int min_time = UINT_MAX;
-        unsigned int delta_time;
-
-        // Find next event across all tracks.
-        for (i = 0; i < song.num_tracks; ++i)
+        if (!song.tracks[i].end_of_track)
         {
-            if (!song.tracks[i].end_of_track)
+            unsigned int time = song.tracks[i].elapsed_time
+                                + MIDI_GetDeltaTime(song.tracks[i].iter);
```

Comment:
Maybe we should make all the time related variables `uint64_t` to be safe. Same with the `midi_track_t` and `midi_song_t` structures.

Reply:
I don't think it is necessary, `elapsed_time` and `delta_time` are in MIDI ticks, `int32_t` should be sufficient.